packages/flutter/test/widgets/binding_attach_root_widget_test.dart

```diff
@@ -9,13 +9,15 @@ import 'package:flutter_test/flutter_test.dart';
 
 void main() {
   test('attachRootWidget will schedule a frame', () async {
-    final WidgetsFlutterBinding binding = WidgetsFlutterBinding.ensureInitialized() as WidgetsFlutterBinding;
+    final WidgetsFlutterBindingWithTestBinaryMessenger binding = WidgetsFlutterBindingWithTestBinaryMessenger();
     expect(SchedulerBinding.instance.hasScheduledFrame, isFalse);
     // Framework starts with detached statue. Sends resumed signal to enable frame.
     final ByteData message = const StringCodec().encodeMessage('AppLifecycleState.resumed')!;
-    await ServicesBinding.instance.defaultBinaryMessenger.handlePlatformMessage('flutter/lifecycle', message, (_) { });
+    await binding.defaultBinaryMessenger.handlePlatformMessage('flutter/lifecycle', message, (_) { });
 
     binding.attachRootWidget(const Placeholder());
     expect(SchedulerBinding.instance.hasScheduledFrame, isTrue);
   });
 }
+
+class WidgetsFlutterBindingWithTestBinaryMessenger extends WidgetsFlutterBinding with TestDefaultBinaryMessengerBinding { }
```

Comment:
Why do we need this new class instead of just using `AutoTestWidgetsFlutterBinding.ensureInitialized()` to create one as we've done in other tests?

Reply:
The test as written specifically wants to test the frame scheduling behavior of `WidgetsFlutterBinding`, not the scheduling behavior in test mode, which is slightly different. It does want to pretend-send a message from the engine, though, for which it needs the TestBinaryMessenger. Hence, this class.

Reply:
Huh. That sounds a bit... complex. Alright, thanks for clarifying!